### ISSUE TITLE
No support for setting a Host header via API (probably should surface requester API endpoint)

### ISSUE THREAD

**Issue by az-catch** [zaproxy/zaproxy#8220]
### Is your feature request related to a problem? Please describe.

Have tried using scripts and replacer to set a `host` header and neither works, both mechanisms get overridden. It looks like support was added in requester, but there's no apparent API support for it.

### Describe the solution you'd like

I'd be happy regardless of where support for it is added as long as it's documented.

### Describe alternatives you've considered

- Replacer Rules
- Scripts

### Would you like to help fix this issue?

- [ ] Yes

**Comment by thc202** [comment]
Just to clarify, do you mean the ZAP API?

Both scripts (e.g. https://github.com/zaproxy/zaproxy/issues/1318#issuecomment-1386984742) and replacer (#5475) allow to send with a custom host header. (Replacer was not yet released to marketplace hough.)

**Comment by thc202** [comment]
The latest version of Replacer add-on supports that: https://www.zaproxy.org/docs/desktop/addons/replacer/#host-header

**Comment by az-catch** [comment]
**Re: Requester**

I mean the HTTP API - i.e. https://www.zaproxy.org/docs/api/

**Re: replacer/scripts**

Neither replacer rules or scripts work - I get these versions FYI:

plugin/replacer-release-15.zap
plugin/scripts-release-43.zap

**With replacer**

I added a rule like this (where `$ZAP_S` is the server and `$ZAP_A`  is an auth string):

`curl -s -X GET $ZAP_S/JSON/replacer/action/addRule/?description=host_replacer\&matchType=REQ_HEADER\&matchRegex=false\&matchString=Host\&replacement=$DOMAIN\&enabled=true\&$ZAP_A`

The rule gets added, but has no effect.

**With scripts**

Host header is applied at the time and will output correctly, but the actual request is not made with the applied host header. Other headers stick. As an example - I have a script that looks like this:

```
// compatibility between nashorn and rhino script
if (typeof println !== 'undefined' && typeof println !== 'function' ) {
    var print = function (str) { println(str); }
}else{
    var println = function (str) { print(str); }
}

function sendingRequest(msg, initiator, helper) {
    print('Sending Request:')
    print('URL: ' + msg.getRequestHeader().getURI().toString())
    msg.getRequestHeader().setHeader('X-Server-Env-PSK', 'xxx');
    msg.getRequestHeader().setHeader('Host', 'override.domain.com');
    msg.getRequestHeader().setHeader('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.3')
    print('request headers: ' + msg.getRequestHeader().toString());
    return msg;
}

function responseReceived(msg, initiator, helper) {
    print('Response Recived:')
    print('URL: ' + msg.getRequestHeader().getURI().toString())
    print('Request Headers:')
    print(msg.getRequestHeader().toString());
    print('Response:')
    print(msg.getResponseHeader().toString())
    print(msg.getResponseBody().toString())
}
```

When requests are made I get output like this (which looks correct):

```
Sending Request:
URL: https://scan.domain.com
request headers: GET https://scan.domain.com HTTP/1.1
Host: override.domain.com
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36
pragma: no-cache
cache-control: no-cache
X-Server-Env-PSK: xxx
```

But when inspecting the request headers in the response it looks like the below. I know this isn't some artifact because the site I am scanning throws a 502 when the host headers are not set correctly and the behavior is consistent if doing a manual request using curl:

```
Request Headers:
GET https://scan.domain.com HTTP/1.1
host: scan.domain.com
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36
pragma: no-cache
cache-control: no-cache
X-Server-Env-PSK: xxx
```

I noted the request headers available at the time of response always have a lower case `host` even thought I am setting it as `Host`, other headers casing is not changed. I've tried using both `Host` and `host` when setting the headers and it makes no difference.

**Comment by kingthorin** [comment]
Please try with Replacer 16 which was just released in the last few days.

**Comment by thc202** [comment]
It was released today :) https://github.com/zaproxy/zaproxy/issues/8220#issuecomment-1833550627

https://github.com/zaproxy/zaproxy/issues/8220#issuecomment-1834442032

You are not configuring the message to override the host header (that was mentioned in the linked comment). But, you can do so with e.g.:
```js
msg.setUserObject({"host.normalization": false});
```

**Comment by thc202** [comment]
No, it's not yet possible to override the host through the ZAP API.
